### ISSUE TITLE
[MCHECKSTYLE-186] FileTabCharacter check not working

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-checkstyle-plugin#213]
**[Dipti Desai](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=maudish)** opened **[MCHECKSTYLE-186](https://issues.apache.org/jira/browse/MCHECKSTYLE-186?redirect=false)** and commented

The FileTabCharacter check doesnt seem to work. Below is my config:

```xml
<module name="Checker">
..
..
<!-- No TAB characters in the source code -->
<module name="FileTabCharacter">
    <property name="eachLine" value="true" />
    <property name="fileExtensions" value="java,xml" />
</module>
..
..
<module name="TreeWalker">
..
..
</module>
</module>
```

I have my xml files - pom.xml and checkstyle config xml containing tabs but none of them are flagged as violations.

Some additional info - my plugin config looks like this:

```xml
<plugin>
    <groupId>org.apache.maven.plugins</groupId>
    <artifactId>maven-checkstyle-plugin</artifactId>
    <version>2.9.1</version>
    <executions>
        <execution>
            <phase>verify</phase>
            <goals>
                <goal>check</goal>
            </goals>
        </execution>
    </executions>
    <configuration>
        <configLocation>checkstyle/kepler-checkstyle-config.xml</configLocation>
        <suppressionsLocation>${project.parent.basedir}${file.separator}checkstyle/kepler-checkstyle-suppressions.xml</suppressionsLocation>
    </configuration>
</plugin>
```

---

**Issue Links:**
- [MCHECKSTYLE-184](https://issues.apache.org/jira/browse/MCHECKSTYLE-184) Allow src/main/resources to be searched by Checkstyle

**Comment by jira-importer** [comment]
**[Dennis Lundberg](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dennisl@apache.org)** commented

Version 2.9.1 of maven-checkstyle-plugin only checks *.java files, unless you change the configuration. See linked issue.

**Comment by jira-importer** [comment]
**[Dipti Desai](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=maudish)** commented

Thanks for your reply. I went through the linked issue and the revisions mentioned there. I may be missing something out here but I didn't really find anything specific that could validate my xml files. Could you please elaborate a little on what could fix this issue?

**Comment by jira-importer** [comment]
**[Dennis Lundberg](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dennisl@apache.org)** commented

How have you configured \<sourceDirectory>? I'm guessing you haven't.
Where are your xml-files?
Are they in the ${project.build.sourceDirectory} directory? I'm guessing no.
If not, then you need to configure the \<sourceDirectory> as well.

**Comment by jira-importer** [comment]
**[Dipti Desai](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=maudish)** commented

`@Dennis`, have you had a chance to look into my query above?

**Comment by jira-importer** [comment]
**[Dipti Desai](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=maudish)** commented

This has been resolved. You may close this. Thanks for all your help.
I am not sure what exactly fixed the problem but I poked around a bit in the maven checkstyle code and got the exact default expression for includes and excludes and modified and used that in my POMs. Also, upgraded to 2.10.

**Comment by jira-importer** [comment]
**[Amichai Rothman](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=amichai)** commented

`@Dipti`, would you be able to share your working configuration? I've been struggling with this for hours as well, with various forms of includes/sourceDirectory/fileExtensions.

Eventually I peeked at the source for DefaultCheckstyleExecutor.java, from which it appears that using the checkstyle-aggregate goal pretty much ignores the configured sourceDirectory. I switched to using the checkstyle goal and it finally picked up the xml files (under resources) and found the violations there.

I'm not entirely sure how the sourceDirectory was desgined to function in aggregate, I'd assumed that I can configure it once in the parent (aggregator) and it would apply to the child modules and then aggregate the result, but the code seems to just take the standard build source directory for each child module, and ignoring the configured sourceDirectory when adding the files to process. If someone can confirm that this is indeed a bug, I'd be happy to open a separate issue for it. If the current behavior is by design, it would be useful to mention this in the documentation, seeing that it may not be intuitive for everyone.

In addition, as logic and comments in the code say, it would be nice if the currently half-hacked resources directory treatment would be made generic: ideally, there would be a separate configuration for the sourcesDirectory, testSourcesDirectory and resourcesDirectory, and each one would have its own includes clause as well - for maximum configurability.

**Comment by jira-importer** [comment]
**[Dipti Desai](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=maudish)** commented

This is what I am currently using in my checkstyle-config.xml
Hope this helps.

```
<!-- No TAB characters in the source code -->
<module name="FileTabCharacter">
    <property name="eachLine" value="true"/>
    <property name="fileExtensions" value="java,xml"/>
</module>
```

**Comment by jira-importer** [comment]
**[Amichai Rothman](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=amichai)** commented

`@Dipti`, and the checkstyle plugin configuration (in the pom)?

**Comment by jira-importer** [comment]
**[Dipti Desai](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=maudish)** commented

```
<plugin>
    <groupId>org.apache.maven.plugins</groupId>
    <artifactId>maven-checkstyle-plugin</artifactId>
    <version>2.10</version>
    <executions>
        <execution>
            <phase>verify</phase>
            <goals>
                <goal>check</goal>
            </goals>
        </execution>
    </executions>
    <configuration>
        <includes>**\/*.xml,**\/*.java</includes>
        <excludes>**/target/**/*,**/test/**/*</excludes>
        <configLocation>checkstyle/kepler-checkstyle-config.xml</configLocation>
        <suppressionsLocation>${project.parent.basedir}${file.separator}checkstyle/kepler-checkstyle-suppressions.xml</suppressionsLocation>
    </configuration>
</plugin>
```

**Comment by jira-importer** [comment]
**[Amichai Rothman](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=amichai)** commented

So no sourceDirectory... interesting. Thanks!
